go/storage/mkvs/db/badger/migrate.go

```diff
@@ -665,15 +681,647 @@ func (v4 *v4Migrator) Migrate() (rversion uint64, rerr error) {
+		return 0, fmt.Errorf("mkvs: failed to unmarshal right hash: %w", err)
+	}
+	pos += hash.Size
+
+	if leftHash.IsEmpty() {
+		n.Left = nil
+	} else {
+		n.Left = &node.Pointer{Clean: true, Hash: leftHash}
+	}
+
+	if rightHash.IsEmpty() {
+		n.Right = nil
+	} else {
+		n.Right = &node.Pointer{Clean: true, Hash: rightHash}
+	}
+
+	n.UpdateHash()
+
+	n.Clean = true
+
+	return pos, nil
+}
+
+func (v5 *v5Migrator) unmarshalV4LeafNode(n *node.LeafNode, data []byte) (int, error) {
+	if len(data) < 1+v4NodeVersionSize+node.DepthSize+node.ValueLengthSize || data[0] != node.PrefixLeafNode {
+		return 0, node.ErrMalformedNode
+	}
+
+	pos := 1
+	// Skip version field.
+	pos += v4NodeVersionSize
+
+	var key node.Key
+	keySize, err := key.SizedUnmarshalBinary(data[pos:])
+	if err != nil {
+		return 0, err
+	}
+	pos += keySize
+	if pos+node.ValueLengthSize > len(data) {
+		return 0, node.ErrMalformedNode
+	}
+
+	valueSize := int(binary.LittleEndian.Uint32(data[pos : pos+node.ValueLengthSize]))
+	pos += node.ValueLengthSize
+	if pos+valueSize > len(data) {
+		return 0, node.ErrMalformedNode
+	}
+
+	value := make([]byte, valueSize)
+	copy(value, data[pos:pos+valueSize])
+	pos += valueSize
+
+	n.Clean = true
+	n.Key = key
+	n.Value = value
+
+	n.UpdateHash()
+
+	return pos, nil
+}
+
+func (v5 *v5Migrator) migrateNode(h hash.Hash, version uint64) (*hash.Hash, error) {
+	item, err := v5.readTxn.Get(v4NodeKeyFmt.Encode(&h))
+	if err != nil {
+		return nil, fmt.Errorf("error loading node %s: %w", h, err)
+	}
+	raw, err := item.ValueCopy(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error loading node %s: %w", h, err)
+	}
+
+	// Decode node type.
+	var (
+		newHash hash.Hash
+		newNode []byte
+	)
+	switch raw[0] {
+	case node.PrefixInternalNode:
+		var n node.InternalNode
+		_, err = v5.unmarshalV4InternalNode(&n, raw)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding internal node %s: %w", h, err)
+		}
+
+		// Migrate children.
+		for _, child := range []*node.Pointer{n.Left, n.Right} {
+			if child == nil {
+				continue
+			}
+			var nh *hash.Hash
+			nh, err = v5.migrateNode(child.Hash, version)
+			if err != nil {
+				return nil, err
+			}
+			child.Hash = *nh
+		}
+		n.UpdateHash()
+		newHash = n.GetHash()
+		newNode, _ = n.MarshalBinary()
+	case node.PrefixLeafNode:
+		var n node.LeafNode
+		_, err = v5.unmarshalV4LeafNode(&n, raw)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding leaf node %s: %w", h, err)
+		}
+		newHash = n.GetHash()
+		newNode, _ = n.MarshalBinary()
+	default:
+		return nil, fmt.Errorf("unknown node type for node %s: 0x%02x", h, raw[0])
+	}
+
+	// Store node. We store the node as if it was created in the version we are migrating regardless
+	// of its original version. This is because we need to be purging any old versions.
+	entry := badger.NewEntry(v5NodeKeyFmt.Encode(&newHash), newNode)
+	err = v5.changeBatch.SetEntryAt(entry, item.Version())
```

Comment:
Shouldn't this set the entry at `version`  (according to the above comment).

Reply:
No, that was the initial thought but the comment is now wrong since we actually also migrate all versions that come after the last finalized version.